libraries/chain/finality/qc.cpp

```diff
@@ -21,6 +21,77 @@ inline std::vector<uint32_t> bitset_to_vector(const vote_bitset_t& bs) {
    return r;
 }
 
+// returns true if vote indicated by active_vote_index in active_policy
+// is the same as vote indicated by pending_vote_index in pending_policy
+bool qc_t::vote_same_at(uint32_t active_vote_index, uint32_t pending_vote_index) const {
+   assert(pending_policy_sig);
+   return active_policy_sig.vote_same_at(*pending_policy_sig,
+                                         active_vote_index,
+                                         pending_vote_index);
+}
+
+// returns true if vote indicated by my_vote_index in strong_votes or
+// weak_votes are is the same as the one indicated by other_vote_index
+// in `other` signature's strong_votes or weak_votes
+bool qc_sig_t::vote_same_at(const qc_sig_t& other, uint32_t my_vote_index, uint32_t other_vote_index) const {
+   auto is_vote_same = [&] (std::optional<vote_bitset_t> my_votes,
+                            uint32_t my_vote_index,
+                            std::optional<vote_bitset_t> other_votes,
+                            uint32_t other_vote_index) -> bool {
+      // both votes must have the same has_value()
+      assert(my_votes.has_value() == other_votes.has_value());
+      // validate indexes
+      assert(my_vote_index < my_votes->size() && other_vote_index < other_votes->size());
+
+      return (*my_votes)[my_vote_index] == (*other_votes)[other_vote_index];
+   };
+
+   // if strong_votes exist, both votes the same
+   if (strong_votes) {
+      return is_vote_same(strong_votes, my_vote_index, other.strong_votes, other_vote_index);
+   }
+
+   // if weak_votes exist, both votes the same
+   if (weak_votes) {
+      return is_vote_same(weak_votes, my_vote_index, other.weak_votes, other_vote_index);
+   }
+
+   assert(true); // should never happen
+   return false; // avoid compile warning
+}
+
+void qc_sig_t::verify_vote_format(const finalizer_policy_ptr& fin_policy) const {
+   assert(fin_policy);
+
+   const auto& finalizers = fin_policy->finalizers;
+   auto num_finalizers = finalizers.size();
+
+   // verify number of finalizers matches with vote bitset size
+   if (strong_votes) {
+      EOS_ASSERT( num_finalizers == strong_votes->size(), invalid_qc_claim,
+                  "vote bitset size is not the same as the number of finalizers for the policy it refers to, "
+                  "vote bitset size: ${s}, num of finalizers for the policy: ${n}",
+                  ("s", strong_votes->size())("n", num_finalizers) );
+   }
+   if (weak_votes) {
+      EOS_ASSERT( num_finalizers == weak_votes->size(), invalid_qc_claim,
+                  "vote bitset size is not the same as the number of finalizers for the policy it refers to, "
+                  "vote bitset size: ${s}, num of finalizers for the policy: ${n}",
+                  ("s", weak_votes->size())("n", num_finalizers) );
+   }
+
+   // verify a finalizer cannot vote both strong and weak
+   if (strong_votes && weak_votes) {
```

Comment:
Need an `EOS_ASSERT` that either `strong_votes` or `weak_votes` is present.

Reply:
https://github.com/AntelopeIO/spring/pull/475/commits/58edf7588ed6bafa7a709d0e8c866d6d33f803e6